credentials/credentials_ext_test.go

```diff
@@ -0,0 +1,338 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package credentials_test
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"log"
+	"net"
+	"testing"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/internal/stubserver"
+	testgrpc "google.golang.org/grpc/interop/grpc_testing"
+	testpb "google.golang.org/grpc/interop/grpc_testing"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+	"google.golang.org/grpc/testdata"
+)
+
+var cert tls.Certificate
+var creds credentials.TransportCredentials
+
+func init() {
+	var err error
+	cert, err = tls.LoadX509KeyPair(testdata.Path("x509/server1_cert.pem"), testdata.Path("x509/server1_key.pem"))
+	if err != nil {
+		log.Fatalf("failed to load key pair: %s", err)
+	}
+	creds, err = credentials.NewClientTLSFromFile(testdata.Path("x509/server_ca_cert.pem"), "x.test.example.com")
+	if err != nil {
+		log.Fatalf("Failed to create credentials %v", err)
+	}
+}
+
+func authorityChecker(ctx context.Context, expectedAuthority string) (*testpb.Empty, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, status.Error(codes.InvalidArgument, "failed to parse metadata")
+	}
+	auths, ok := md[":authority"]
+	if !ok {
+		return nil, status.Error(codes.InvalidArgument, "no authority header")
+	}
+	if len(auths) != 1 {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("no authority header, auths = %v", auths))
+	}
+	if auths[0] != expectedAuthority {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("invalid authority header %v, expected %v", auths[0], expectedAuthority))
+	}
+	return &testpb.Empty{}, nil
+}
+
+func checkUnavailableRPCError(t *testing.T, err error) {
+	t.Helper()
+	if err == nil {
+		t.Fatalf("EmptyCall() should fail")
+	}
+	s, ok := status.FromError(err)
+	if !ok {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if s.Code() != codes.Unavailable {
+		t.Fatalf("EmptyCall() = _, %v, want _, error code: %v", s.Code(), codes.Unavailable)
+	}
+}
+
+// Tests the grpc.CallAuthority option with TLS credentials. This test verifies
+// that the provided authority is correctly propagated to the server when using TLS.
+// It covers both positive and negative cases:  correct authority and incorrect
+// authority, expecting the RPC to fail with `UNAVAILABLE` status code error in
+// the latter case.
+func (s) TestAuthorityCallOptionsWithTLSCreds(t *testing.T) {
+	tests := []struct {
+		name         string
+		expectedAuth string
+		wantRPCError bool
+	}{
+		{
+			name:         "CorrectAuthorityWithTLS",
+			expectedAuth: "auth.test.example.com",
+			wantRPCError: false,
```

Comment:
What if you simplified this by making this field `wantStatus`. So, in this subtest the field will be set to `codes.Ok` and the next one will have `codes.Unavailable`. And you wont need the `checkUnavailableRPCError` helper function at all.

The last line can be rewritten as
client := testgrpc.NewTestServiceClient(cc)
if _, err :=client.EmptyCall(ctx, &testpb.Empty{}, grpc.CallAuthority(tt.expectedAuth)); status.Code(err) != tt.wantStatus {
  t.Fatalf("EmptyCall returned status: %v, want %v", status.Code(err), tt.wantStatus)
}